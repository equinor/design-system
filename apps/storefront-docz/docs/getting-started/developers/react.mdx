---
title: Developers
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
