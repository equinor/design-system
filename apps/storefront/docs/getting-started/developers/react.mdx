---
title: Developers
metaTitle: Developers react
tabs:
  - Introduction
  - CSS framework
  - React
  - Figma
mode: draft
route: /getting-started/developers/react/
---

:warning: Information coming soon
