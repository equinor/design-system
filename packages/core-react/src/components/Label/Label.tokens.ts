import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  colors,
  spacings: { comfortable },
  typography,
} = tokens

type LabelToken = ComponentToken

export const label: LabelToken = {
  background: colors.ui.background__light.rgba,
  typography: {
    ...typography.input.label,
    color: colors.text.static_icons__tertiary.rgba,
  },
  spacings: {
    left: comfortable.small,
    right: comfortable.small,
    top: '6px',
    bottom: '6px',
  },
  states: {
    disabled: {
      typography: {
        color: colors.interactive.disabled__text.rgba,
      },
    },
  },
}
