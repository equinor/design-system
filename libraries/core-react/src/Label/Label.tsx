import * as React from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '../_common/templates'
import { label as tokens } from './Label.tokens'

const LabelBase = styled.label`
  display: flex;
  justify-content: space-between;
  position: relative;

  ${typographyTemplate(tokens.typography)}
  margin-left: ${tokens.spacings.left};
  margin-right: ${tokens.spacings.right};
  color: ${tokens.color};
`

const Text = styled.span`
  margin: 0;
`

type TextfieldProps = {
  label: string
  meta: string
  inputId: string
}

const TextFieldLabel = React.forwardRef<HTMLLabelElement, TextfieldProps>(
  function TextFieldLabel(props, ref) {
    const { label = '', meta = '', inputId } = props

    return (
      <LabelBase ref={ref} htmlFor={inputId}>
        <Text>{label}</Text>
        <Text>{meta}</Text>
      </LabelBase>
    )
  },
)

// Label.displayName = 'eds-text-field-label'

export { TextFieldLabel as Label }
