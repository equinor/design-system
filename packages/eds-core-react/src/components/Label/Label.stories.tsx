import { Label, LabelProps, Input } from '../..'
import { StoryFn, Meta } from '@storybook/react'
import page from './Label.docs.mdx'

const meta: Meta<typeof Label> = {
  title: 'Inputs/Label',
  component: Label,
  parameters: {
    docs: {
      page,
    },
  },
}

export default meta

export const Introduction: StoryFn<LabelProps> = (args) => (
  <Label label="I'm a label, play with me!" {...args} />
)

export const WithMeta: StoryFn<LabelProps> = () => (
  <Label label="Speed" meta="km/h" />
)
WithMeta.storyName = 'With meta text'

export const Disabled: StoryFn<LabelProps> = () => (
  <Label
    label="I'm disabled, that means I belong to a disabled input field"
    disabled
  />
)

export const Accessiblity: StoryFn<LabelProps> = () => {
  // To wrap the input component inside the label element is not yet supported
  return (
    <>
      <Label label="I use the htmlFor prop" htmlFor="speed" />
      <Input type="text" id="speed" />
    </>
  )
}
