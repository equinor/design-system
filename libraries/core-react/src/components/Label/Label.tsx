import * as React from 'react'
import { LabelHTMLAttributes } from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '../../utils'
import { label as tokens } from './Label.tokens'

type LabelBaseType = {
  disabledText: boolean
}

const LabelBase = styled.label<LabelBaseType>`
  display: flex;
  justify-content: space-between;
  position: relative;
  ${typographyTemplate(tokens.typography)}
  margin-left: ${tokens.spacings.left};
  margin-right: ${tokens.spacings.right};
  color: ${({ disabledText }) =>
    disabledText ? tokens.disabled.color : tokens.color};
`

const Text = styled.span`
  margin: 0;
`

export type LabelProps = {
  label: string
  meta?: string
  disabled?: boolean
} & LabelHTMLAttributes<HTMLLabelElement>

export const Label = React.forwardRef<HTMLLabelElement, LabelProps>(
  function Label(props, ref) {
    const { label = '', meta, disabled = false, ...other } = props

    return (
      /*  @TODO: Other props spread has to be at the end for downshift to create the for attribute */
      <LabelBase ref={ref} disabledText={disabled} {...other}>
        <Text>{label}</Text>
        {meta && <Text>{meta}</Text>}
      </LabelBase>
    )
  },
)

// Label.displayName = 'eds-text-field-label'
