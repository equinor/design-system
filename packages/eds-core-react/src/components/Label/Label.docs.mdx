import { Links } from './../../../.storybook/components'
import { Primary, Canvas, Controls, Meta } from '@storybook/blocks'
import * as ComponentStories from './Label.stories'

<Meta of={ComponentStories} />

# Label

Is intended to use if you need to more flexibility than the wrapped TextField or select components give.

<Links 
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Label/Label.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Primary />
<Controls />

## Usage

```tsx
import { Label } from '@equinor/eds-core-react'

<Label label="label text" />
```

## Examples

### With `meta` text

<Canvas of={ComponentStories.WithMeta} />

### Disabled

<Canvas of={ComponentStories.Disabled} />

### Accessiblity

It's important to link the `Label` to the corresponding input element. 
Use the React version of the html for attribute htmlFor as `Label` does not support to wrap the input field.

<Canvas of={ComponentStories.Accessiblity} />
