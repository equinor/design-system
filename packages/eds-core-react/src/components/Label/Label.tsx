import { LabelHTMLAttributes, forwardRef } from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '@equinor/eds-utils'
import { label as tokens } from './Label.tokens'

type LabelBaseType = {
  disabledText: boolean
}

const LabelBase = styled.label<LabelBaseType>`
  display: flex;
  justify-content: space-between;
  position: relative;
  ${typographyTemplate(tokens.typography)}
  margin-left: ${tokens.spacings.left};
  margin-right: ${tokens.spacings.right};
  color: ${({ disabledText }) =>
    disabledText
      ? tokens.states.disabled.typography.color
      : tokens.typography.color};
`

const Text = styled.span`
  margin: 0;
`

export type LabelProps = {
  label: string
  meta?: string
  disabled?: boolean
} & LabelHTMLAttributes<HTMLLabelElement>

export const Label = forwardRef<HTMLLabelElement, LabelProps>(
  function Label(props, ref) {
    const { label = '', meta, disabled = false, ...other } = props

    return (
      /*  @TODO: Other props spread has to be at the end for downshift to create the for attribute */
      <LabelBase ref={ref} disabledText={disabled} {...other}>
        <Text>{label}</Text>
        {meta && <Text>{meta}</Text>}
      </LabelBase>
    )
  },
)

// Label.displayName = 'eds-text-field-label'
