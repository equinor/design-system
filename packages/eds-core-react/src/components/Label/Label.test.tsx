/* eslint-disable no-undef */
import { render, cleanup, screen } from '@testing-library/react'
import 'jest-styled-components'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { Label } from './Label'

afterEach(cleanup)

describe('Label', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<Label label="label" />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<Label label="Some label text" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test with meta', async () => {
    const { container } = render(<Label label="Some label" meta="Meta" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test with input', async () => {
    const { container } = render(
      <>
        <Label label="Some label text" htmlFor="input" />
        <input id="input" />
      </>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Has correct label text', () => {
    const labelText = 'Some label'
    render(<Label label={labelText} />)

    expect(screen.getByText(labelText)).toBeInTheDocument()
  })

  it('Can add a meta text', () => {
    const labelText = 'Some label'
    const metaText = 'Meta'
    render(<Label label={labelText} meta={metaText} />)

    expect(screen.getByText(metaText)).toBeInTheDocument()
  })

  const StyledLabel = styled(Label)`
    margin-top: 48px;
  `
  it('Can extend the css of the component', () => {
    const labelText = 'Some label'
    render(<StyledLabel label={labelText} data-testid="label" />)

    expect(screen.getByTestId('label')).toHaveStyleRule('margin-top', '48px')
  })
})
