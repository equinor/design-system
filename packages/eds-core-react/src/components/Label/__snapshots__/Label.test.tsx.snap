// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  justify-content: space-between;
  position: relative;
  margin: 0;
  color: var(--eds_text__static_icons__tertiary, rgba(111, 111, 111, 1));
  font-family: Equinor;
  font-size: 0.750rem;
  font-weight: 500;
  line-height: 1.333em;
  text-align: left;
  margin-left: 8px;
  margin-right: 8px;
  color: var(--eds_text__static_icons__tertiary, rgba(111, 111, 111, 1));
}

.c1 {
  margin: 0;
}

<label
    class="c0"
  >
    <span
      class="c1"
    >
      label
    </span>
  </label>
</DocumentFragment>
`;
